File: libs/OscFramework_2.0/Source Code/Samples/OscDemo/CS/Receiver/Program.cs
using System;
using System.Net;
using System.Collections.Generic;
using Bespoke.Common;
using Bespoke.Common.Osc;

namespace Receiver
{
    public enum DemoType
    {
        Udp,
        Tcp,
        Multicast
    }

	public class Program
	{
		public static void Main(string[] args)
		{
            OscServer oscServer;
            DemoType demoType = GetDemoType();
            switch (demoType)
            {
                case DemoType.Udp:
                    oscServer = new OscServer(TransportType.Udp, IPAddress.Loopback, Port);
                    break;

                case DemoType.Tcp:
                    oscServer = new OscServer(TransportType.Tcp, IPAddress.Loopback, Port);
                    break;

                case DemoType.Multicast:
                    oscServer = new OscServer(IPAddress.Parse("224.25.26.27"), Port);
                    break;

                default:
                    throw new Exception("Unsupported receiver type.");
            }
            
            oscServer.FilterRegisteredMethods = false;
			oscServer.RegisterMethod(AliveMethod);
            oscServer.RegisterMethod(TestMethod);
            oscServer.BundleReceived += new EventHandler<OscBundleReceivedEventArgs>(oscServer_BundleReceived);
			oscServer.MessageReceived += new EventHandler<OscMessageReceivedEventArgs>(oscServer_MessageReceived);
            oscServer.ReceiveErrored += new EventHandler<ExceptionEventArgs>(oscServer_ReceiveErrored);
            oscServer.ConsumeParsingExceptions = false;

            oscServer.Start();

			Console.WriteLine("Osc Receiver: " + demoType.ToString());
			Console.WriteLine("Press any key to exit.");
			Console.ReadKey();
			oscServer.Stop();
		}

        private static DemoType GetDemoType()
        {
            Dictionary<ConsoleKey, DemoType> keyMappings = new Dictionary<ConsoleKey, DemoType>();
            keyMappings.Add(ConsoleKey.D1, DemoType.Udp);
            keyMappings.Add(ConsoleKey.D2, DemoType.Tcp);
            keyMappings.Add(ConsoleKey.D3, DemoType.Multicast);

            Console.WriteLine("\nWelcome to the Bespoke Osc Receiver Demo.\nPlease select the type of receiver you would like to use:");
            Console.WriteLine("  1. Udp\n  2. Tcp\n  3. Udp Multicast");

            ConsoleKeyInfo key = Console.ReadKey();
            while (keyMappings.ContainsKey(key.Key) == false)
            {
                Console.WriteLine("\nInvalid selection\n");
                Console.WriteLine("  1. Udp\n  2. Tcp\n  3. Udp Multicast");
                key = Console.ReadKey();
            }

            Console.Clear();

            return keyMappings[key.Key];
        }

        private static void oscServer_BundleReceived(object sender, OscBundleReceivedEventArgs e)
        {
            sBundlesReceivedCount++;

            OscBundle bundle = e.Bundle;
            Console.WriteLine(string.Format("\nBundle Received [{0}:{1}]: Nested Bundles: {2} Nested Messages: {3}", bundle.SourceEndPoint.Address, bundle.TimeStamp, bundle.Bundles.Count, bundle.Messages.Count));
            Console.WriteLine("Total Bundles Received: {0}", sBundlesReceivedCount);
        }

		private static void oscServer_MessageReceived(object sender, OscMessageReceivedEventArgs e)
		{
            sMessagesReceivedCount++;

            OscMessage message = e.Message;

            Console.WriteLine(string.Format("\nMessage Received [{0}]: {1}", message.SourceEndPoint.Address, message.Address));
            Console.WriteLine(string.Format("Message contains {0} objects.", message.Data.Count));

            for (int i = 0; i < message.Data.Count; i++)
            {
                string dataString;

                if (message.Data[i] == null)
                {
                    dataString = "Nil";
                }
                else
                {
                    dataString = (message.Data[i] is byte[] ? BitConverter.ToString((byte[])message.Data[i]) : message.Data[i].ToString());
                }
                Console.WriteLine(string.Format("[{0}]: {1}", i, dataString));
            }

            Console.WriteLine("Total Messages Received: {0}", sMessagesReceivedCount);
		}

        private static void oscServer_ReceiveErrored(object sender, ExceptionEventArgs e)
        {
            Console.WriteLine("Error during reception of packet: {0}", e.Exception.Message);
        }

		private static readonly int Port = 5103;
		private static readonly string AliveMethod = "/osctest/alive";
        private static readonly string TestMethod = "/osctest/test";

        private static int sBundlesReceivedCount;
        private static int sMessagesReceivedCount;
	}
}

File: libs/OscFramework_2.0/Source Code/Samples/OscDemo/VB/Receiver/Program.vb
Module Program
    Public Enum DemoType
        Udp
        Tcp
        Multicast
    End Enum

    Sub Main()
        Dim demoType As DemoType = GetDemoType()

        Select Case demoType
            Case Program.DemoType.Udp
                sOscServer = New OscServer(TransportType.Udp, IPAddress.Loopback, Port)

            Case Program.DemoType.Tcp
                sOscServer = New OscServer(TransportType.Tcp, IPAddress.Loopback, Port)

            Case Program.DemoType.Multicast
                sOscServer = New OscServer(IPAddress.Parse("224.25.26.27"), Port)

            Case Else
                Throw New Exception("Unsupported receiver type.")
        End Select

        sOscServer.FilterRegisteredMethods = False
        sOscServer.RegisterMethod(AliveMethod)
        sOscServer.RegisterMethod(TestMethod)
        sOscServer.ConsumeParsingExceptions = False

        sOscServer.Start()

        Console.WriteLine("Osc Receiver: " + demoType.ToString())
        Console.WriteLine("Press any key to exit.")
        Console.ReadKey()

        sOscServer.Stop()
    End Sub

    Private Sub sOscServer_BundleReceived(ByVal sender As Object, ByVal e As OscBundleReceivedEventArgs) Handles sOscServer.BundleReceived
        sBundlesReceivedCount += 1

        Dim bundle As OscBundle = e.Bundle

        Console.WriteLine(String.Format("{0}Bundle Received [{1}:{2}]: Nested Bundles: {3} Nested Messages: {4}", Environment.NewLine, bundle.SourceEndPoint.Address, bundle.TimeStamp, bundle.Bundles.Count, bundle.Messages.Count))
        Console.WriteLine("Total Bundles Received: {0}", sBundlesReceivedCount)
    End Sub

    Private Sub sOscServer_MessageReceived(ByVal sender As Object, ByVal e As OscMessageReceivedEventArgs) Handles sOscServer.MessageReceived
        sMessagesReceivedCount += 1

        Dim message As OscMessage = e.Message

        Console.WriteLine(String.Format("{0}Message Received [{1}]: {2}", Environment.NewLine, message.SourceEndPoint.Address, message.Address))
        Console.WriteLine(String.Format("Message contains {0} objects.", message.Data.Count))

        For i As Integer = 0 To message.Data.Count - 1
            Dim dataString As String

            If message.Data(i) Is Nothing Then
                dataString = "Nil"
            Else
                dataString = If(TypeOf message.Data(i) Is Byte(), BitConverter.ToString(message.Data(i)), message.Data(i).ToString())
            End If

            Console.WriteLine(String.Format("[{0}]: {1}", i, dataString))
        Next

        Console.WriteLine("Total Messages Received: {0}", sMessagesReceivedCount)
    End Sub

    Private Sub sOscServer_ReceiveErrored(ByVal sender As Object, ByVal e As ExceptionEventArgs) Handles sOscServer.ReceiveErrored
        Console.WriteLine("Error during reception of packet: {0}", e.Exception.Message)
    End Sub

    Private Function GetDemoType() As DemoType
        Dim keyMappings As Dictionary(Of ConsoleKey, DemoType) = New Dictionary(Of ConsoleKey, DemoType)()
        keyMappings.Add(ConsoleKey.D1, DemoType.Udp)
        keyMappings.Add(ConsoleKey.D2, DemoType.Tcp)
        keyMappings.Add(ConsoleKey.D3, DemoType.Multicast)

        Console.WriteLine("{0}Welcome to the Bespoke Osc Receiver Demo.{0}Please select the type of receiver you would like to use:", Environment.NewLine)
        Console.WriteLine("  1. Udp{0}  2. Tcp{0}  3. Udp Multicast", Environment.NewLine)

        Dim key As ConsoleKeyInfo = Console.ReadKey()
        While keyMappings.ContainsKey(key.Key) = False
            Console.WriteLine("{0}Invalid selection{0}", Environment.NewLine)
            Console.WriteLine("  1. Udp{0}  2. Tcp{0}  3. Udp Multicast", Environment.NewLine)
            key = Console.ReadKey()
        End While

        Console.Clear()

        Return keyMappings(key.Key)
    End Function

    Private ReadOnly Port As Integer = 5103

    Private ReadOnly AliveMethod As String = "/osctest/alive"
    Private ReadOnly TestMethod As String = "/osctest/test"

    Private WithEvents sOscServer As OscServer
    Private sBundlesReceivedCount As Integer
    Private sMessagesReceivedCount As Integer
End Module
